metals/src/main/scala/scala/meta/internal/metals/TastyHandler.scala

```diff
@@ -0,0 +1,151 @@
+      tastyResponse.map { tasty =>
+        val command = new l.ExecuteCommandParams(
+          "metals-show-tasty",
+          List[Object](tasty).asJava
+        )
+        languageClient.metalsExecuteClientCommand(command)
+        scribe.debug(s"Executing show TASTy ${command}")
+      }
+    } else {
+      (httpServer(), parseJsonParams(params)) match {
+        case (Some(server), Some(uri)) =>
+          Future.successful(
+            Urls.openBrowser(server.address + s"/tasty?file=$uri")
+          )
+        case (None, _) =>
+          Future.successful {
+            scribe.warn(
+              "Unable to run show tasty. Make sure `isHttpEnabled` is set to `true`."
+            )
+          }
+        case other =>
+          scribe.error(s"Show TASTy command failed $other")
+          languageClient.showMessage(
+            Messages.showTastyFailed
+          )
+          Future.successful(())
+      }
+    }
+
+  def getTastyForURI(uri: URI): Future[String] =
+    getTasty(
+      uri,
+      clientConfig.isCommandInHtmlSupported(),
+      clientConfig.isHttpEnabled()
+    )
+
+  private def getTasty(
+      params: l.ExecuteCommandParams
+  ): Future[String] =
+    parseJsonParams(params) match {
+      case Some(path) =>
+        val uri = new URI(path)
+        getTastyForURI(uri)
+      case None =>
+        val error = s"Error, invalid show TASTy arguments $params"
+        scribe.error(error)
+        Future.successful(error)
+    }
+
+  private def parseJsonParams(
+      commandParams: l.ExecuteCommandParams
+  ): Option[String] = for {
+    args <- Option(commandParams.getArguments)
+    arg <- args.asScala.headOption.collect {
+      case js: JsonPrimitive if js.isString => js
+    }
+  } yield arg.getAsString()
+
+  /**
+   * For given uri (could be both .scala and .tasty file) try to find:
+   * - build target
+   * - uri to existing tasty file
+   * Iff they both exists get decoded tasty file content (pc.getTasty takes into account if client supports html)
+   */
+  private def getTasty(
+      uri: URI,
+      isHtmlSupported: Boolean,
+      isHttpEnabled: Boolean
+  ): Future[String] = {
+    val absolutePathOpt = Try(AbsolutePath.fromAbsoluteUri(uri)) match {
+      case Success(value) => Some(value)
+      case Failure(exception) =>
+        scribe.error(exception.toString())
+        None
+    }
+    val pcAndTargetUriOpt = for {
+      absolutePath <- absolutePathOpt
+      buildTargetId <- buildTargets.inverseSources(absolutePath)
+      buildTarget <- buildTargets.scalaTarget(buildTargetId)
+      pc <- compilers.loadCompilerForTarget(buildTarget)
+      tastyFileURI <- getTastyFileURI(
+        uri,
+        buildTarget.scalac.getClassDirectory.toAbsolutePath
+      )
+    } yield (pc, tastyFileURI)
+
+    pcAndTargetUriOpt match {
+      case Some((pc, tastyUri)) =>
+        pc.getTasty(tastyUri, isHtmlSupported, isHttpEnabled).asScala
+      case _ => Future.successful("Error, .tasty file doesn't exist")
```

Comment:
Do you think it's possible to maybe change this up a bit and not return just a string as an argument, but rather something more structured? Right now if this happens (which it did to me when testing), this will need to be handled differently by the client (especially when they aren't using html) than the actual tasty string that is returned. The only way to detect this currently would is to basically do a string.startsWith("Error") type of check. 
Even something simple like:

```ts
interface TastyReponse {
  error?: String
  body?: String
}
```
Then it'd be possible to check for an error. Either that, or don't return `successful` in this case. wdyt?

Reply:
Yeah, I cannot agree more. 
I even tried to create discriminated union type but I don't think it's possible due to how command serialization works :( so I ended with sth like you proposed.